wnfs/examples/file_variants.rs

```diff
@@ -0,0 +1,63 @@
+//! This example shows how to store multiple byte arrays per file, by storing
+//! additional data in a file's metadata, which links out to externally encrypted data.
+
+use anyhow::Result;
+use chrono::Utc;
+use rand::thread_rng;
+use std::rc::Rc;
+use wnfs::private::{
+    forest::{hamt::HamtForest, traits::PrivateForest},
+    PrivateFile, PrivateForestContent,
+};
+use wnfs_common::MemoryBlockStore;
+
+#[async_std::main]
+async fn main() -> Result<()> {
+    // The usual in-memory testing setup for WNFS
+    let store = &MemoryBlockStore::default();
+    let rng = &mut thread_rng();
+    let forest = &mut HamtForest::new_rsa_2048(rng);
+
+    // Create a new file (detached from any directory)
+    let mut file = Rc::new(
+        PrivateFile::with_content(
+            &forest.empty_name(),
+            Utc::now(),
+            b"main content".to_vec(),
+            forest,
+            store,
+            rng,
+        )
+        .await?,
+    );
+
+    // Create some content that's stored encrypted in the private forest.
+    // The PrivateForestContent struct holds the keys and pointers to look it back up.
+    // We use the file's name as the "path" for this content. This means anyone
+    // who had write access to the file will have write access to the external content.
+    let content = PrivateForestContent::new(
+        file.header.get_name(),
+        b"secondary content".to_vec(),
+        forest,
+        store,
+        rng,
+    )
+    .await?;
+
+    // We store the content in the file metadata.
+    // This will update the `file: Rc<PrivateFile>` for us with a new reference.
+    content.store_in_metadata(file.get_metadata_mut_rc()?, "thumbnail")?;
```

Comment:
Following the suggestion above, this would then read more naturally as:

```
file.upsert_metadata("thumbnail", |_| content.as_metadata_value());
```

What do you think?

Reply:
I like the idea of an `as_metadata_value` function :)

I don't want to re-introduce functions like `upsert_metadata` that really should belong to the `Metadata` implementation, and partially exists there already (`put`, `get`, `delete`, etc.).

It's somewhat annoying at the moment that sometimes you have a `&mut PrivateFile` and sometimes you have a `&mut Rc<PrivateFile>` (and we don't expose `prepare_next_revision`), but :man_shrugging: not sure about a better solution there in general.

Reply:
You are right. I can't come up with a fine solution for this.